nexus/db-queries/src/db/datastore/region.rs

```diff
@@ -159,67 +162,95 @@ impl DataStore {
+                        let datasets = diesel::delete(region_dsl::region)
+                            .filter(region_dsl::id.eq_any(region_ids))
+                            .returning(region_dsl::dataset_id)
+                            .get_results::<Uuid>(conn)?;
+
+                        // Update datasets to which the regions belonged.
+                        for dataset in datasets {
+                            let dataset_total_occupied_size: Option<
+                                diesel::pg::data_types::PgNumeric,
+                            > = region_dsl::region
+                                .filter(region_dsl::dataset_id.eq(dataset))
+                                .select(diesel::dsl::sum(
+                                    region_dsl::block_size
+                                        * region_dsl::blocks_per_extent
+                                        * region_dsl::extent_count,
+                                ))
+                                .nullable()
+                                .get_result(conn)?;
+
+                            let dataset_total_occupied_size: i64 = if let Some(
+                                dataset_total_occupied_size,
+                            ) =
+                                dataset_total_occupied_size
+                            {
+                                let dataset_total_occupied_size: db::model::ByteCount =
+                                    dataset_total_occupied_size.try_into().map_err(
+                                        |e: anyhow::Error| {
+                                            TxnError::CustomError(
+                                                RegionDeleteError::NumericError(
+                                                    e.to_string(),
+                                                ),
+                                            )
+                                        },
+                                    )?;
+
+                                dataset_total_occupied_size.into()
+                            } else {
+                                0
+                            };
+
+                            diesel::update(dataset_dsl::dataset)
+                                .filter(dataset_dsl::id.eq(dataset))
+                                .set(
+                                    dataset_dsl::size_used
+                                        .eq(dataset_total_occupied_size),
+                                )
+                                .execute(conn)?;
+                        }
+
+                        Ok(())
+                    })
+                    .await
+                    .map_err(|e: TxnError| {
+                        if e.retry_transaction() {
+                            BackoffError::transient(Error::internal_error(
+                                &format!("Retryable transaction error {:?}", e)
+                            ))
```

Comment:
This might be more of a philosophical question, but....

... should we use this pattern for all transactions? I struggle to understand why **this specific transaction** should be treated differently from others.

Reply:
I think we should retry all transactions that CRDB tells us to :) but I think we'll need to convince ourselves that this won't create unbounded work for Nexus.

Reply:
Would you be okay making a separate issue for this? After this PR merges, I could help with this.

Reply:
Definitely, opened https://github.com/oxidecomputer/omicron/issues/3814